File: Longest ZigZag Path in a Binary Tree.java
class Solution {

public class Pair{
    int forward=-1;
    int backward=-1;
    int max=0;
}
public Pair lon(TreeNode root){
    if(root==null) return new Pair();
    Pair left=lon(root.left);
    Pair right=lon(root.right);
    Pair ans=new Pair();
    ans.max=Math.max(Math.max(left.max,right.max),Math.max(left.backward,right.forward)+1);
    ans.forward=left.backward+1;
    ans.backward=right.forward+1;
    return ans;
}
public int longestZigZag(TreeNode root) {
   
    Pair p=lon(root);
    return p.max;
    
}
}

File: Longest ZigZag Path in a Binary Tree.cpp
class Solution {
public:

    void dfs(TreeNode* root, int pleft, int pright,int &ans){
        if(root==NULL) return ;
        if(root->right){
            ans = max(ans, pleft+1);
            dfs(root->right, 0, pleft+1,ans);
        }
        if(root->left){
            ans = max(ans, pright+1);
            dfs(root->left, pright+1, 0,ans);
            
        }
    }
    int longestZigZag(TreeNode* root) {
    int ans=0;
        dfs(root, 0, 0,ans);
        return ans;
    }
};
